clippy_lints/src/use_self.rs

```diff
@@ -318,3 +323,35 @@ fn lint_path_to_variant(cx: &LateContext<'_>, path: &Path<'_>) {
         span_lint(cx, span);
     }
 }
+
+/// Returns `true` if types `a` and `b` have the same lifetime parameters, otherwise returns
+/// `false`.
+///
+/// This function does not check that types `a` and `b` are the same types.
+fn same_lifetimes<'tcx>(a: MiddleTy<'tcx>, b: MiddleTy<'tcx>) -> bool {
+    use rustc_middle::ty::{Adt, GenericArgKind};
+    match (&a.kind(), &b.kind()) {
+        (&Adt(_, args_a), &Adt(_, args_b)) => {
+            args_a
+                .iter()
+                .zip(args_b.iter())
+                .all(|(arg_a, arg_b)| match (arg_a.unpack(), arg_b.unpack()) {
+                    (GenericArgKind::Lifetime(inner_a), GenericArgKind::Lifetime(inner_b)) => inner_a == inner_b,
+                    (GenericArgKind::Type(type_a), GenericArgKind::Type(type_b)) => same_lifetimes(type_a, type_b),
+                    _ => true,
+                })
+        },
+        _ => a == b,
+    }
+}
+
+/// Returns `true` if `ty` has no lifetime parameter, otherwise returns false.
+fn has_no_lifetime(ty: MiddleTy<'_>) -> bool {
+    use rustc_middle::ty::{Adt, GenericArgKind};
+    match ty.kind() {
+        &Adt(_, args) => !args
+            .iter()
+            .any(|arg| matches!(arg.unpack(), GenericArgKind::Lifetime(..))),
```

Comment:
Should this take into account in case a lifetime is `infer`? Kinda the same as having no lifetime I suppose?

Reply:
I have no idea about this. There are already tests about this case and they do pass. I think namely of [this line in the tests](https://github.com/rust-lang/rust-clippy/blob/00ff8c92d395d08325301c1a1462fc38abd82053/tests/ui/use_self.rs#L64).

Unfortunately, I wish I were more knowledgeable about the subject to really understand how that works, but I only had the tests to try and understand it. I know not the difference between having no lifetime parameter and having an `Infer` lifetime. I would be happy to have input on it.

The same goes for your previous comment about `Infer`.

Reply:
`infer` is `_` in short. Maybe try adding a test returning a `Foo<'_>` to see if you cover this case?

Reply:
As of now
```rs
fn f(&self) -> Foo<'_> {
    *self
}
```
Does not trigger the lint. Should it?

Reply:
I think that would make a false positive of
```rs
fn eq<'b>(&self, other: Foo<'b>) -> bool {
    let x: Foo<'_> = other;
    self.foo_str == x
}
```

Reply:
Sounds like it should no?

Reply:
Can be done in a follow-up.

Reply:
Replacing `Foo<'_>` in my example makes a `Self` (`Foo<'a>`) of `other`, which is a `Foo<'b>`. That feels wrong to me.

Reply:
I only had in mind the case of the returned value. Maybe we could only check it for `-> Foo<'_>`?

Reply:
Thinking about it some more, that might not be this trivial. Definitely worth opening an issue to handle these cases and leave it for a later PR.

Reply:
For now, we can just ignore inferred lifetimes. Please add a TODO comment to the code :)